microprofile/WFLY-14798-upgrade-reactive-messaging-2.0.adoc

```diff
@@ -0,0 +1,143 @@
+    Emitter<String> emitter;
+
+    @Inject @Channel("my-stream")
+    Publisher<String> dest;
+
+    @POST
+    public PublisherBuilder<String> publish(@FormParam("value") String value) {
+        return emitter.send(value);
+    }
+
+    @GET
+    public Publisher poll(String value) {
+        return dest;
+    }
+}
+----
+Sending data via the Emitter works fine, but as we will see in the Non-Requirements section there are some issues with returning the `Publisher` via JAX-RS in the `poll()` method.
+
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/EAP7-1659
+* https://issues.redhat.com/browse/WFLY-14798
+
+=== Related Issues
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* TBD
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE.
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+We will upgrade to the SmallRye version implementing MicroProfile Reactive Messaging 2.0.
+
+All other code changes to support the changes will happen in WildFly itself. Tests for the new functionality will be moved over from the https://github.com/wildfly-extras/wildfly-mp-reactive-feature-pack[WildFly Extras MP Reactive Feature Pack] where this work was incubated.
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+For community all installation types are relevant. For product, this is currently targeted for EAP XP.
+
+== Requirements
+=== Hard Requirements
+It must be possible to:
+
+* Provision a server with MicroProfile Reactive Messaging for Kafka. The layers will be the same as in the https://issues.redhat.com/browse/WFLY-13640[original implementation].
+* Deploy a Reactive Messaging application into WildFly
+* Interact with Reactive Messaging in-memory and Kafka streams via the Reactive Messaging 2.0 APIS
+** MicroProfile Config will be used to configure the connections to the Kafka streams.
+* Get metrics for messages sent via the streams
+** The microprofile-reactive-messaging layer will introduce a dependency on the `microprofile-metrics` layer, and the `microprofile-reactive-messaging-smallrye` subsystem will require the capabilities provided by the `microprofile-metrics-smallrye` subsystem.
+* If a user tries to use annotations coming from SmallRye Reactive Messaging API jar that are not in the Reactive Messaging 2.0 specification, reasonable effort is made to raise an error on deployment.
+
+=== Nice-to-Have Requirements
+
+It might be possible to make the dependency on the `microprofile-metrics` layer, and the `microprofile-metrics-smallrye` subsystem capabilities, optional.
```

Comment:
I don't really know if there is enough time to do this for this release. Perhaps we could create a follow-up RFE?

Reply:
If this isn't implemented please do file a follow-up JIRA. Over time we want to drive users toward Micrometer instead of MP Metrics so we don't want anything forcing people to use MP Metrics.

Reply:
Actually, this seems to be optional already - https://github.com/wildfly/wildfly/pull/14475#discussion_r694866377 all that is needed is the metrics API jar